src/terminal/parser/OutputStateMachineEngine.cpp

```diff
@@ -724,6 +726,16 @@ bool OutputStateMachineEngine::ActionOscDispatch(const wchar_t /*wch*/,
             fSuccess = _dispatch->SetColorTableEntry(tableIndex, dwColor);
             TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCCT);
             break;
+
+        case OscActionCodes::SetForegroundColor:
+            fSuccess = _dispatch->SetDefaultForeground(dwColor);
+            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCCT);

```

Comment:
These two should probably get their own entries in `TermTelemetry`, instead of overloading `TermTelemetry::Codes::OSCCT`

Reply:
Added OSCFG and OSCBG to the enum and added them to the WriteFinalTraceLog -- is that all there is?

Reply:
Yep!

Reply:
@zadjii-msft I was poking through OutputStateMachineEngine.cpp, and it looks like this code does not actually do anything yet, since the first switch (on line 689 in the main branch as of today) reads:

```
    switch (sOscParam)
    {
    case OscActionCodes::SetIconAndWindowTitle:
    case OscActionCodes::SetWindowIcon:
    case OscActionCodes::SetWindowTitle:
        fSuccess = _GetOscTitle(pwchOscStringBuffer, cchOscString, &pwchTitle, &sCchTitleLength);
        break;
    case OscActionCodes::SetColor:
        fSuccess = _GetOscSetColorTable(pwchOscStringBuffer, cchOscString, &tableIndex, &dwColor);
        break;
    case OscActionCodes::SetForegroundColor:
    case OscActionCodes::SetBackgroundColor:
    case OscActionCodes::SetCursorColor:
        fSuccess = _GetOscSetColor(pwchOscStringBuffer, cchOscString, &dwColor);
        break;
    case OscActionCodes::ResetCursorColor:
        // the console uses 0xffffffff as an "invalid color" value
        dwColor = 0xffffffff;
        fSuccess = true;
        break;
    default:
        // If no functions to call, overall dispatch was a failure.
        fSuccess = false;
        break;
    }
```
Since `fSuccess` is initialized as `false`, this seems like it makes the next switch, which is wrapped in an `if` depending on `fSuccess` skip the OSC 10 and 11 cases:
```
    if (fSuccess)
    {
        switch (sOscParam)
        {
        case OscActionCodes::SetIconAndWindowTitle:
        case OscActionCodes::SetWindowIcon:
        case OscActionCodes::SetWindowTitle:
            fSuccess = _dispatch->SetWindowTitle({ pwchTitle, sCchTitleLength });
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCWT);
            break;
        case OscActionCodes::SetColor:
            fSuccess = _dispatch->SetColorTableEntry(tableIndex, dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCCT);
            break;
        case OscActionCodes::SetForegroundColor:
            fSuccess = _dispatch->SetDefaultForeground(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCFG);
            break;
        case OscActionCodes::SetBackgroundColor:
            fSuccess = _dispatch->SetDefaultBackground(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCBG);
            break;
        case OscActionCodes::SetCursorColor:
            fSuccess = _dispatch->SetCursorColor(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCSCC);
            break;
        case OscActionCodes::ResetCursorColor:
            fSuccess = _dispatch->SetCursorColor(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCRCC);
            break;
        default:
            // If no functions to call, overall dispatch was a failure.
            fSuccess = false;
            break;
        }
    }
```

Doesn't this mean that the code in this PR right now isn't actually doing anything, since it's always skipped over?  Should `fSuccess` instead be initialized as `true`?

Reply:
The case additions [here](https://github.com/microsoft/terminal/pull/891/files#diff-6ddb7535cb6d02274d60e614e3780e4dR700) should make sure we _do_ dispatch to `_GetOscSetColor`, if I'm reading it properly.

Reply:
@DHowett Sorry if I'm being dense, but when you initialize the variable `fsuccess` for the method, you initialize it as false, rather than true.  Therefore, those case additions don't help. See line 683 in the current master branch.

Reply:
https://github.com/microsoft/terminal/blob/94bcbb92045f9daf16168cd82b1abdf731065bc8/src/terminal/parser/OutputStateMachineEngine.cpp#L683-L703

The switch on 689 has a case at 699,700 that covers these Osc codes, and that changes `fSuccess`.

Sure is! Select a line or block of lines (with shift), click the `...` and choose _copy permalink_. Just pasting that URL will make github insert an inline code block.

![image](https://user-images.githubusercontent.com/14316954/65824299-6354b180-e21b-11e9-9641-08a0b5d49c57.png)

![image](https://user-images.githubusercontent.com/14316954/65824300-66e83880-e21b-11e9-9f5a-e3bda5ce74ff.png)

Reply:
Ok, sorry, one last question, since I'm hoping to try to implement another OSC code (52) in the future:  Why does that switch case automagically set `fSuccess` to true without having to add
```
fsuccess = true;
break;
```

Reply:
`fSuccess = _GetOscSetColor(...)`, and `_GetOscSetColor` returns whether _it_ succeeded.

Reply:
In the future, you’ll wonder why the entire exchange ultimately does _fail_. Getting a false out of the last step in the dispatch loop means “and also send it over the pseudoconsole, because I didn’t handle it to a state that is complete enough for a connected terminal.” That’s why you will see some “if in VtIo mode, return false” exchanges.

Reply:
@DHowett Ohhhh, I think I see, those cases get combined in C++ if not followed by a break.  I assumed there was always an implicit break between cases, but it's actually working like a goto and will execute all operations until the next break. Good to know! Thanks!